src/Elf.cpp

```diff
@@ -0,0 +1,989 @@
+
+template <int bits>
+struct Types;
+
+template <>
+struct Types<32> {
+    typedef uint32_t addr_t;
+    typedef int32_t addr_off_t;
+};
+
+template <typename T>
+struct Ehdr {
+    typedef typename T::addr_t addr_t;
+    typedef typename T::addr_off_t addr_off_t;
+
+    uint8_t e_ident[16];
+    uint16_t e_type;
+    uint16_t e_machine;
+    uint32_t e_version;
+    addr_t e_entry;
+    addr_t e_phoff;
+    addr_t e_shoff;
+    uint32_t e_flags;
+    uint16_t e_ehsize;
+    uint16_t e_phentsize;
+    uint16_t e_phnum;
+    uint16_t e_shentsize;
+    uint16_t e_shnum;
+    uint16_t e_shstrndx;
+};
+
+
+template <typename T>
+struct Phdr {
+    typedef typename T::addr_t addr_t;
+    typedef typename T::addr_off_t addr_off_t;
+
+    uint32_t p_type;
+    uint32_t p_offset;
+    addr_t p_vaddr;
+    addr_t p_paddr;
+    uint32_t p_filesz;
+    uint32_t p_memsz;
+    uint32_t p_flags;
+    uint32_t p_align;
+};
+
+template <typename T>
+struct Shdr {
+    typedef typename T::addr_t addr_t;
+    typedef typename T::addr_off_t addr_off_t;
+
+    uint32_t sh_name;
+    uint32_t sh_type;
+    addr_t sh_flags;
+    addr_t sh_addr;
+    addr_t sh_offset;
+    addr_t sh_size;
+    uint32_t sh_link;
+    uint32_t sh_info;
+    addr_t sh_addralign;
+    addr_t sh_entsize;
+};
+
+template <typename T>
+struct Rel {
+    typedef typename T::addr_t addr_t;
+    typedef typename T::addr_off_t addr_off_t;
+
+    addr_t r_offset;
+    addr_t r_info;
+
+    uint32_t r_type() const {
+        if (sizeof(addr_t) == 8) {
+            return r_info & 0xffffffff;
+        } else {
+            return r_info & 0xff;
+        }
+    }
+
+    uint32_t r_sym() const {
+        if (sizeof(addr_t) == 8) {
+            return (uint64_t)r_info >> 32;
+        } else {
+            return r_info >> 8;
+        }
+    }
+
+    static addr_t make_info(uint32_t type, uint32_t sym) {
+        if (sizeof(addr_t) == 8) {
+            return (uint64_t)type | ((uint64_t)sym << 32);
+        } else {
+            return (type & 0xff) | (sym << 8);
+        }
+    }
+
+    void set_r_type(uint32_t type) {
+        r_info = make_info(type, r_sym());
+    }
+
+    void set_r_sym(uint32_t sym) {
+        r_info = make_info(r_type(), sym);
+    }
+
+    Rel(addr_t offset, addr_t info)
+        : r_offset(offset), r_info(info) {}
+
+    Rel(addr_t offset, uint32_t type, uint32_t sym)
+        : r_offset(offset), r_info(make_info(type, sym)) {}
+};
+
+template <typename T>
+struct Rela : public Rel<T> {
+    typedef typename T::addr_t addr_t;
+    typedef typename T::addr_off_t addr_off_t;
+
+    addr_off_t r_addend;
+
+    Rela(addr_t offset, addr_t info, addr_off_t addend)
+        : Rel<T>(offset, info), r_addend(addend) {}
+
+    Rela(addr_t offset, uint32_t type, uint32_t sym, addr_off_t addend)
+        : Rel<T>(offset, type, sym), r_addend(addend) {}
+};
+
+template <typename T>
+struct Sym;
+
+template <>
+struct Sym<Types<32>> {
```

Comment:
It's a specialization, I haven't bothered to define the 64 bit version yet. Will add a comment.